coders/heic.c

```diff
@@ -299,6 +341,26 @@ static Image *ReadHEICImage(const ImageInfo *image_info,
   heif_image_release(heif_image);
   heif_image_handle_release(image_handle);
   heif_context_free(heif_context);
+
+  /*
+    There is a discrepancy between EXIF data and the actual orientation of
+    image pixels. ReadImage processes "exif:Orientation" expecting pixels to be
+    oriented accordingly. However, in HEIF the pixels are NOT rotated.
+
+    There are two solutions to this problem: either reset the EXIF Orientation tag
+    so it matches the orientation of pixels, or rotate the pixels to match EXIF data.
+   */
+  option=GetImageOption(image_info,"heic:preserve-orientation");
+  if (IsStringTrue(option) == MagickTrue)
+    {
+      Image
+        *original_image = image;
+      image = CompensateOrientation(original_image, exception);
+      DestroyImageList(original_image);
```

Comment:
You should handle this inside the `CompensateOrientation` method. You are now destroying the `image` when `exif:Orientation` is not set.

Reply:
Oh, true. Fill fix promptly.

Reply:
Pushed.
I am sorry for the silly mistake